### PR TITLE
Add Hybrid scan page pruning when offset index is absent

### PR DESCRIPTION
**Pull request by mhaseeb123**
## Description

This PR enables the hybrid scan reader to still prune data pages after page header decode (save decompression and decode) using the row mask when offset index is not present.

Note that list column pages cannot be pruned in this fallback method as list rows may spill across page boundaries when offset index is absent.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/examples/hybrid_scan_io/hybrid_scan_composer.cpp [hunk 1]
```diff
@@ -426,22 +426,6 @@ std::unique_ptr<cudf::table> hybrid_scan(
   }
 }
 
-// Specialization for two-step read without page index
```
Comment (mhaseeb123) [inline cpp/examples/hybrid_scan_io/hybrid_scan_composer.cpp:433]: This case is now handled so enable

File: cpp/src/io/parquet/experimental/hybrid_scan_chunking.cu [hunk 2]
```diff
@@ -40,8 +40,16 @@ void hybrid_scan_reader_impl::handle_chunking(
...
+    auto const data_page_mask_pghdr = [&]() {
+      if (not _has_offset_index and not _row_mask.is_empty()) {
+        return compute_data_page_mask_with_page_headers();
+      }
+      return thrust::host_vector<bool>(data_page_mask.begin(), data_page_mask.end());
+    }();
+
     // Must be called as soon as we create the pass
-    set_pass_page_mask(data_page_mask);
+    set_pass_page_mask(data_page_mask_pghdr.empty() ? data_page_mask : data_page_mask_pghdr);
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/hybrid_scan_chunking.cu]: Use fallback page mask computation if needed

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp [hunk 3]
```diff
@@ -1239,4 +1247,68 @@ void hybrid_scan_reader_impl::set_pass_page_mask(std::span<bool const> data_page
                "Encountered mismatch in number of pass pages and page mask size");
 }
 
+thrust::host_vector<bool> hybrid_scan_reader_impl::compute_data_page_mask_with_page_headers()
+{
+  auto& pass = *_pass_itm_data;
+  pass.pages.device_to_host_async(_stream);
+  _stream.sync();
+
+  std::vector<cudf::size_type> page_row_offsets;
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp:1487]: This function is actually simpler than it looks. We are essentially doing the same thing as in `_extended_metadata->compute_data_page_mask()`. Here's the rundown:

Go over all pages and:
- dict page: not needed since we want a data page mask.
- data page of list col: push -1 to `row_range_map` meaning we will inject a `true` in the final mask for it. (See comment on L1272)
- data page: first page in the chunk, push start row and end row, otherwise just push end row to `page_row_offsets`.

Call the `compute_row_range_selection_mask` to get a row rang mask and gather the final data page mask using it and the `row_range_map`

File: cpp/src/io/parquet/experimental/page_index_filter.cu [hunk 4]
```diff
@@ -976,6 +976,68 @@ std::unique_ptr<cudf::column> aggregate_reader_metadata::build_row_mask_with_pag
...
 }
 
+thrust::host_vector<bool> compute_row_range_selection_mask(
+  cudf::column_view const& row_mask,
+  cudf::size_type row_mask_offset,
+  cudf::size_type total_rows,
+  std::span<cudf::size_type const> page_row_offsets,
+  cudf::size_type max_page_size,
+  cuda::stream_ref stream)
+{
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/page_index_filter.cu]: This helper is literally just moved version of code from LHS. See the big red block on lhs in `compute_data_page_mask`. We just call this helper from `compute_data_page_mask`. This is done so we can call this helper from `compute_data_page_mask_with_page_headers()` function you just saw above in `hybrid_scan_impl.cpp`

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp [hunk 5]
```diff
@@ -1441,6 +1449,70 @@ void hybrid_scan_reader_impl::set_pass_page_mask(std::span<bool const> data_page
                "Encountered mismatch in number of pass pages and page mask size");
 }
 
+thrust::host_vector<bool> hybrid_scan_reader_impl::compute_data_page_mask_with_page_headers()
```
Comment (vuule) [inline cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp]: unverified AI correctness complaint:
The fallback drops the null handling that the offset-index path has. `aggregate_reader_metadata::compute_data_page_mask()` forces `row_mask[i] = true` for null (undetermined) rows before searching, and also returns early when every row in the pass is null. `compute_data_page_mask_with_page_headers()` searches the raw `bool` data buffer, so a null row whose underlying data byte is `false` prunes its page and its rows are silently lost. This is reachable for filter columns: `build_row_mask_with_page_index()` only requires column+offset indexes for the *stats-participating* columns, while `_has_offset_index` is computed over all filter columns, so a nullable row mask can coexist with `_has_offset_index == false`. Either replicate the null fixup (the mask is a `mutable_column_view` on the filter path) or refuse the fallback with a `CUDF_EXPECTS` when `_row_mask.nullable() and null_count() > 0`.

File: cpp/include/cudf/io/experimental/hybrid_scan.hpp [hunk 6]
```diff
@@ -656,7 +656,7 @@ class hybrid_scan_reader {
     std::size_t chunk_read_limit,
     std::size_t pass_read_limit,
     std::span<size_type const> row_group_indices,
-    cudf::column_view const& row_mask,
+    cudf::mutable_column_view const& row_mask,
```
Comment (mhaseeb123) [inline cpp/include/cudf/io/experimental/hybrid_scan.hpp]: Breaking change: Make this mutable now

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp [hunk 7]
```diff
@@ -696,8 +697,9 @@ table_with_metadata hybrid_scan_reader_impl::materialize_filter_columns(
 
   auto data_page_mask = thrust::host_vector<bool>{};
   if (mask_data_pages == use_data_page_mask::YES) {
+    _row_mask      = resolve_row_mask_nulls(row_mask, stream);
     data_page_mask = _extended_metadata->compute_data_page_mask(
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp]: No need to pass in `_row_mask_offset` as its always zero here anyway (reset by `prepare_materialization`)

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp [hunk 8]
```diff
@@ -696,8 +697,9 @@ table_with_metadata hybrid_scan_reader_impl::materialize_filter_columns(
 
   auto data_page_mask = thrust::host_vector<bool>{};
   if (mask_data_pages == use_data_page_mask::YES) {
+    _row_mask      = set_nulls_to_true(row_mask, stream);
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp]: Set nulls to true (keep row) for mutable row masks passed to filter column materializers. Payload column materializers only take in non-nullable row masks

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp [hunk 9]
```diff
@@ -805,8 +808,9 @@ void hybrid_scan_reader_impl::setup_chunking_for_filter_columns(
...
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp]: Set nulls to true (keep row) for mutable row masks passed to filter column materializers. Payload column materializers only take in non-nullable row masks

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp (hunk 5 above)
Comment (mhaseeb123) [reply cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp]: Fixed by explicitly setting nulls to true (keep row) for mutable row masks passed to filter column materializers. Payload column materializers only take in non-nullable row masks so we are all good in that case

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.hpp [hunk 10]
```diff
@@ -253,7 +253,7 @@ class hybrid_scan_reader_impl : public parquet::detail::reader_impl {
     std::size_t chunk_read_limit,
     std::size_t pass_read_limit,
     std::span<std::vector<size_type> const> row_group_indices,
-    cudf::column_view const& row_mask,
+    cudf::mutable_column_view const& row_mask,
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/hybrid_scan_impl.hpp]: Take in mutable column view

File: cpp/src/io/parquet/experimental/page_index_filter.cu [hunk 11]
```diff
@@ -606,234 +600,6 @@ struct page_stats_to_row_mask_converter : public page_stats_caster {
   }
 };
 
-/*
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/page_index_filter.cu:624]: Moved as is to `page_index_filter_utils.cu` in anon namespace as it's only needed locally in there. Cleans up this TU

File: cpp/src/io/parquet/experimental/page_index_filter.cu [hunk 12]
```diff
@@ -978,12 +744,10 @@ std::unique_ptr<cudf::column> aggregate_reader_metadata::build_row_mask_with_pag
     page_stats_table, stats_expr.get_stats_expr().get(), stream, mr);
 }
 
-template <typename ColumnView>
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/page_index_filter.cu:987]: No need to handle mutable column view here anymore or even the row_mask_offset

File: cpp/src/io/parquet/experimental/page_index_filter.cu [hunk 13]
```diff
@@ -1099,112 +863,26 @@ thrust::host_vector<bool> aggregate_reader_metadata::compute_data_page_mask(
     });
   }
 
-  // Make sure all row_mask elements contain valid values even if they are nulls
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/page_index_filter.cu:1108]: Already taken care of by `set_nulls_to_true` function

File: cpp/src/io/parquet/experimental/page_index_filter.cu [hunk 14]
```diff
@@ -1099,112 +863,26 @@ thrust::host_vector<bool> aggregate_reader_metadata::compute_data_page_mask(
...
-                       });
-    }
-  } else {
-    CUDF_EXPECTS(not row_mask.nullable() or row_mask.null_count() == 0,
-                 "Row mask must not contain nulls for payload columns");
-  }
-
-  auto const mr = cudf::get_current_device_resource_ref();
-
-  // Compute fenwick tree level offsets and total size (level 1 and higher)
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/page_index_filter.cu:1126]: All this gone inside the `compute_row_range_selection_mask` function on the RHS as is

File: cpp/src/io/parquet/experimental/page_index_filter_utils.cu [hunk 15]
```diff
@@ -5,23 +5,290 @@
...
 
 #include <algorithm>
 #include <numeric>
 #include <vector>
 
 namespace cudf::io::parquet::experimental::detail {
 
+namespace {
+
+/*
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/page_index_filter_utils.cu:63]: Moved from `page_index_filter.cu` as is.

File: cpp/src/io/parquet/experimental/page_index_filter_utils.cu [hunk 16]
```diff
@@ -158,24 +425,87 @@ rmm::device_uvector<size_type> compute_page_indices_async(
...
+}
 
-  while (current_level_size > 0) {
-    auto const block_size = 1 << current_level;
-    if (std::cmp_greater(block_size, max_page_size)) { break; }
-    tree_level_offsets.push_back(tree_level_offsets.back() + current_level_size);
-    current_level_size =
-      current_level_size == 1 ? 0 : cudf::util::div_rounding_up_safe(current_level_size, 2);
-    current_level++;
+thrust::host_vector<bool> compute_row_range_selection_mask(
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/page_index_filter_utils.cu:476]: Moved as is from `page_index_filter.cu`  - the fenwick tree build and query part

File: cpp/src/io/parquet/experimental/page_index_filter.cu [hunk 17]
```diff
@@ -998,15 +762,15 @@ thrust::host_vector<bool> aggregate_reader_metadata::compute_data_page_mask(
...
-  // Return an empty vector if all rows are invalid or all rows are required
-  if (std::cmp_equal(row_mask.null_count(row_mask_offset, row_mask_offset + total_rows, stream),
-                     total_rows) or
-      cudf::detail::all_of(row_mask.template begin<bool>() + row_mask_offset,
-                           row_mask.template begin<bool>() + row_mask_offset + total_rows,
+  // Return an empty vector if all rows are required
+  if (cudf::detail::all_of(row_mask.begin<bool>(),
+                           row_mask.begin<bool>() + total_rows,
                            cuda::std::identity{},
                            stream)) {
```
Comment (vuule) [inline cpp/src/io/parquet/experimental/page_index_filter.cu]: `compute_row_range_selection_mask` checks this as well, can be safely removed here.

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp [hunk 18]
```diff
@@ -1441,6 +1449,72 @@ void hybrid_scan_reader_impl::set_pass_page_mask(std::span<bool const> data_page
                "Encountered mismatch in number of pass pages and page mask size");
 }
 
+thrust::host_vector<bool> hybrid_scan_reader_impl::compute_data_page_mask_with_page_headers()
+{
+  auto& pass = *_pass_itm_data;
+  pass.pages.device_to_host_async(_stream);
+  _stream.sync();
```
Comment (vuule) [inline cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp]: this is currently redundant, since `decode_page_headers_impl` does the same at its tail, and pages are not updated between the two points.

File: cpp/src/io/parquet/experimental/page_index_filter.cu
Comment (vuule) [inline cpp/src/io/parquet/experimental/page_index_filter.cu]: This is a bit of an overstatement now

**Review by vuule — COMMENTED** [review]
few minor comments

File: cpp/src/io/parquet/experimental/page_index_filter.cu (hunk 17 above)
Comment (mhaseeb123) [reply cpp/src/io/parquet/experimental/page_index_filter.cu]: This was more of an optimization (No need to even collect host level info and return). Moved this to a common util and used it here and at the other call site of `compute_row_range_selection_mask`. Removed it from `compute_row_range_selection_mask` altogether

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp (hunk 18 above)
Comment (mhaseeb123) [reply cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp]: Great catch. Fixed in 1fd4b66

File: cpp/src/io/parquet/experimental/page_index_filter.cu
Comment (mhaseeb123) [reply cpp/src/io/parquet/experimental/page_index_filter.cu]: Specified in 1fd4b66

File: java/src/test/java/ai/rapids/cudf/HybridScanReaderTest.java [hunk 19]
```diff
@@ -825,6 +861,49 @@ void testMaterializePayloadColumnsChunkExactTotal(@TempDir Path tmp) throws IOEx
     }
   }
 
+  /**
+   * Verifies the chunked payload pipeline drains the same 19,999 rows when page pruning
+   * falls back to page-header row counts on a file with no page index.
+   */
+  @Test
+  void testMaterializePayloadColumnsChunkPagePruningWithoutPageIndex(@TempDir Path tmp)
```
Comment (pmattione-nvidia) [inline java/src/test/java/ai/rapids/cudf/HybridScanReaderTest.java:870]: Add a test to cover lists to make sure that they don't prune since rows may span page boundaries.  Though I think (perhaps in a future PR) we could prune payload list pages as well: we know what rows the pages are on after `compute_page_sizes()` is called, so we could skip the value decode (though not the rep/def-level decode).

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp (hunk 8 above)
Comment (pmattione-nvidia) [reply cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp]: wouldn't it be faster to not incur this cost of copying and modifying the mask, but instead make the downstream function interpret nulls as true?

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp [hunk 20]
```diff
@@ -1504,6 +1512,75 @@ void hybrid_scan_reader_impl::set_pass_page_mask(std::span<bool const> data_page
...
+      row_range_map.push_back(-1);
+      continue;
+    }
+
+    auto const page_start = chunk.start_row + page.chunk_row;
+    auto const page_end   = page_start + page.num_rows;
+    max_page_size         = std::max<cudf::size_type>(max_page_size, page_end - page_start);
+
+    // Starting a new column chunk. Push page start row
+    if (previous_chunk_idx == -1 or page.chunk_idx != previous_chunk_idx) {
```
Comment (pmattione-nvidia) [inline cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp]: i would hope only the 2nd condition is necessary?

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp [hunk 21]
```diff
@@ -1504,6 +1512,75 @@ void hybrid_scan_reader_impl::set_pass_page_mask(std::span<bool const> data_page
...
+      page_row_offsets.push_back(page_start);
+      previous_chunk_idx = page.chunk_idx;
+    }
+
+    // Push row range index and page end row
+    row_range_map.push_back(page_row_offsets.size() - 1);
```
Comment (pmattione-nvidia) [inline cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp:1521]: could just do page_row_offsets.back()

File: java/src/test/java/ai/rapids/cudf/HybridScanReaderTest.java (hunk 19 above)
Comment (mhaseeb123) [reply java/src/test/java/ai/rapids/cudf/HybridScanReaderTest.java:870]: Added a test now. List column page pruning in a future PR.

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp (hunk 21 above)
Comment (pmattione-nvidia) [reply cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp:1521]: this didn't change yet

File: cpp/src/io/parquet/experimental/page_index_filter.cu (hunk 13 above)
Comment (pmattione-nvidia) [reply cpp/src/io/parquet/experimental/page_index_filter.cu:1108]: need to update your comment here now that this is gone

File: cpp/src/io/parquet/experimental/page_index_filter_utils.cu [hunk 22]
```diff
@@ -5,23 +5,331 @@
...
 
+namespace {
+
+/**
+ * @brief Functor to read the row mask, i.e. the zeroth Fenwick tree level
+ *
+ * Nulls are read as retained rows here so that pages aren't accidentally pruned due to
+ * unavailable page-level statistics (represented as nulls)
+ */
+struct row_mask_accessor {
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/page_index_filter_utils.cu:39]: Add a row mask accessor that reads `nulls` as `true` and is used to build and query the level0 of the Fenwick tree.

File: cpp/src/io/parquet/experimental/page_index_filter_utils.cu [hunk 23]
```diff
@@ -5,23 +5,331 @@
...
+  cudf::size_type prev_level_size;
+  cudf::size_type current_level_size;
+
+  /**
+   * @brief Reads an element from the previous tree level
+   *
+   * @param prev_level_idx Previous tree level element index
+   * @return Value of the element at the previous tree level
+   */
+  __device__ bool inline read_prev_level(cudf::size_type prev_level_idx) const noexcept
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/page_index_filter_utils.cu:85]: Use the above accessor for the zeroth level, direct access otherwise

File: cpp/src/io/parquet/experimental/page_index_filter_utils.cu [hunk 24]
```diff
@@ -5,23 +5,331 @@
...
+   * @return Boolean indicating if a `true` value is found in the fenwick tree
+   */
+  template <boundary Boundary>
+  __device__ bool inline constexpr query_fenwick_tree(cudf::size_type boundary_pos,
+                                                      cudf::size_type tree_level,
+                                                      cudf::size_type block_size) const noexcept
+  {
+    auto const position = (Boundary == boundary::START) ? boundary_pos : boundary_pos - block_size;
+    auto const mask_index = position >> tree_level;
+    return tree_level == 0 ? row_mask(mask_index) : tree_level_ptrs[tree_level][mask_index];
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/page_index_filter_utils.cu:233]: Use the accessor here for zeroth level

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp (hunk 21 above)
Comment (mhaseeb123) [reply cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp:1521]: I think `page_row_offsets.back()` would be incorrect here. We want to insert the index into the page_row_offsets vector (growing as pages are inserted) not the last element

File: cpp/src/io/parquet/experimental/page_index_filter.cu [hunk 25]
```diff
@@ -998,19 +762,14 @@ thrust::host_vector<bool> aggregate_reader_metadata::compute_data_page_mask(
                std::invalid_argument);
 
   CUDF_EXPECTS(
-    std::cmp_less_equal(row_mask_offset + total_rows, row_mask.size()),
+    std::cmp_equal(total_rows, row_mask.size()),
     "Encountered a mismatch in number of rows in the row group pass and the row mask size",
     std::overflow_error);
+  CUDF_EXPECTS(
```
Comment (pmattione-nvidia) [inline cpp/src/io/parquet/experimental/page_index_filter.cu]: Does this need to be removed now?

Comment (pmattione-nvidia) [reply cpp/src/io/parquet/experimental/page_index_filter.cu]: Since the checks passed I think we need a test that has nulls in the row_mask

Comment (mhaseeb123) [reply cpp/src/io/parquet/experimental/page_index_filter.cu]: Removed the check. Though I don't think adding a test for this is necessary since no hybrid scan API produces a row mask with nulls. Even if so, these nulls are only interpreted as true when pruning filter col pages (we update the row mask with the final state of rows kept and make it non-nullable as we materialize the filter columns). For payload columns we require non-nullable here anyway

Comment (pmattione-nvidia) [reply cpp/src/io/parquet/experimental/page_index_filter.cu]: `build_row_mask_with_page_index_stats` returns a nullable mask and any conjunct against an absent statistic evaluates to null.  E.g. cudf's writer omits min/max for a float or double column containing a NaN, so page stats are absent and the row mask comes back with nulls. `compute_data_page_mask` runs at the top of `materialize_filter_columns` on the caller's mask exactly as handed in, while `update_row_mask` only forces validity at the end, so the nullable mask is what page pruning actually sees.

So the missing test is: write a table with a double column containing a NaN, build the row mask from page-index stats, materialize filter columns with use_data_page_mask::YES, and assert the rows under the null entries survive.

Comment (mhaseeb123) [reply cpp/src/io/parquet/experimental/page_index_filter.cu]: Added in f2ce44d

**Review by pmattione-nvidia — APPROVED** [review]

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.hpp [hunk 26]
```diff
@@ -619,6 +624,11 @@ class hybrid_scan_reader_impl : public parquet::detail::reader_impl {
 
   std::optional<std::vector<std::string>> _filter_columns_names;
 
+  // Non-owning view of the caller's row mask, only valid for the duration of a single
+  // materialization or chunking setup call, during which the pass page mask is computed. Null
+  // entries mean the row could not be pruned and is therefore treated as a surviving row.
+  cudf::column_view _row_mask{};
```
Comment (vuule) [inline cpp/src/io/parquet/experimental/hybrid_scan_impl.hpp]: this is a non-owning view that persists in error cases, ideally we would clear it. not blocking.

File: cpp/src/io/parquet/experimental/page_index_filter_utils.hpp [hunk 27]
```diff
@@ -71,14 +71,32 @@ compute_page_row_offsets_and_colchunk_page_offsets(
   rmm::device_async_resource_ref mr);
 
 /**
- * @brief Computes the offsets of the Fenwick tree levels (level 1 and higher) until the tree level
- * block size becomes larger than the maximum page (search range) size
+ * @brief Checks whether every row is reatained by the boolean row mask
```
Comment (vuule) [inline cpp/src/io/parquet/experimental/page_index_filter_utils.hpp]:
```suggestion
 * @brief Checks whether every row is retained by the boolean row mask
```

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp [hunk 28]
```diff
@@ -731,8 +732,9 @@ table_with_metadata hybrid_scan_reader_impl::materialize_filter_columns(
 
   auto data_page_mask = thrust::host_vector<bool>{};
   if (mask_data_pages == use_data_page_mask::YES) {
+    _row_mask      = row_mask;
     data_page_mask = _extended_metadata->compute_data_page_mask(
```
Comment (vuule) [inline cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp]: In the no-offset-index case the work is done twice: each call site invokes `compute_data_page_mask()`, which validates, runs are_all_rows_retained()`, collects schema indices, then bails out at `page_index_filter.cu:788`, after which `setup_next_pass()` runs `are_all_rows_retained()` again.
Consider computing the data page mask in one place (`setup_next_pass`, which already has both paths in view), or skipping the metadata call when `_has_offset_index` is false.

File: cpp/include/cudf/io/experimental/hybrid_scan.hpp [hunk 29]
```diff
@@ -676,7 +676,7 @@ class hybrid_scan_reader {
    * @param pass_read_limit Limit on the memory used for reading and decompressing data. `0` if
    * there is no limit
    * @param row_group_indices Input row groups indices
-   * @param row_mask Boolean column indicating which rows need to be read
+   * @param[in,out] row_mask Mutable boolean column indicating surviving rows
```
Comment (vuule) [inline cpp/include/cudf/io/experimental/hybrid_scan.hpp]: is this really an out param, isn't row_mask const?

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.hpp [hunk 30]
```diff
@@ -387,6 +387,11 @@ class hybrid_scan_reader_impl : public parquet::detail::reader_impl {
    */
   void set_sparse_pass_page_mask(std::span<cudf::device_span<uint8_t const> const> page_data);
 
+  /**
+   * @brief Compute a data page mask from the decoded page headers.
+   */
+  [[nodiscard]] thrust::host_vector<bool> compute_data_page_mask_with_page_headers();
```
Comment (vuule) [inline cpp/src/io/parquet/experimental/hybrid_scan_impl.hpp]: maybe document the precondition?

File: cpp/src/io/parquet/experimental/page_index_filter_utils.cu [hunk 31]
```diff
@@ -158,24 +465,74 @@ rmm::device_uvector<size_type> compute_page_indices_async(
...
-  cudf::size_type current_level_size = cudf::util::div_rounding_up_safe(level0_size, 2);
-  cudf::size_type current_level      = 1;
+thrust::host_vector<bool> compute_row_range_selection_mask(
+  cudf::column_view const& row_mask,
+  std::span<cudf::size_type const> page_row_offsets,
+  cudf::size_type max_page_size,
+  cuda::stream_ref stream)
+{
+  // Need at least two offsets (or one range) to search the Fenwick tree
+  if (page_row_offsets.size() < 2) return thrust::host_vector<bool>{};
```
Comment (vuule) [inline cpp/src/io/parquet/experimental/page_index_filter_utils.cu]:
```suggestion
  if (page_row_offsets.size() < 2) { return thrust::host_vector<bool>{}; }
```

**Review by vuule — COMMENTED** [review]
some small comments, nothing major

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.hpp (hunk 26 above)
Comment (mhaseeb123) [reply cpp/src/io/parquet/experimental/hybrid_scan_impl.hpp]: Removed this altogether in 3e576ad

File: cpp/src/io/parquet/experimental/page_index_filter_utils.hpp (hunk 27 above)
Comment (mhaseeb123) [reply cpp/src/io/parquet/experimental/page_index_filter_utils.hpp]: Applied in 3e576ad

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp (hunk 28 above)
Comment (mhaseeb123) [reply cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp]: Moved computation of data_page_mask at a central location inside `prepare_data` (when offset index is present along with input row mask) in 3e576ad, or (as existing) inside setup_next_pass (no offset index and input row mask) using page headers.

File: cpp/include/cudf/io/experimental/hybrid_scan.hpp (hunk 29 above)
Comment (mhaseeb123) [reply cpp/include/cudf/io/experimental/hybrid_scan.hpp]: Improved doc in 3e576ad

File: cpp/src/io/parquet/experimental/hybrid_scan_impl.hpp (hunk 30 above)
Comment (mhaseeb123) [reply cpp/src/io/parquet/experimental/hybrid_scan_impl.hpp]: I think it's now clear with the call site in 3e576ad

File: cpp/src/io/parquet/experimental/page_index_filter_utils.cu (hunk 31 above)
Comment (mhaseeb123) [reply cpp/src/io/parquet/experimental/page_index_filter_utils.cu]: Applied in 3e576ad

**Review by vuule — APPROVED** [review]

File: python/pylibcudf/tests/io/test_experimental_hybrid_scan.py [hunk 32]
```diff
@@ -431,6 +431,73 @@ def test_hybrid_scan_materialize_columns(
...
+        plc.gpumemoryview(
+            rmm.DeviceBuffer.to_device(
+                simple_parquet_bytes[r.offset : r.offset + r.size],
+                plc.utils._get_stream(),
+            )
+        )
+        for r in reader.payload_column_chunks_byte_ranges(
+            row_groups, simple_parquet_options
+        )
+    ]
```
Comment (Matt711) [inline python/pylibcudf/tests/io/test_experimental_hybrid_scan.py:469]: We need to keep the source bytes alive since `to_device` is async on the on the non-default stream. If we don't, python can garbage collect it before the GPU's done reading from it.
```suggestion
    # the caller is responsible for keeping the source bytes alive until
    # synchronize_stream() below runs.
    # See https://github.com/rapidsai/rmm/issues/2521
    src_bytes = simple_parquet_bytes[r.offset : r.offset + r.size]
    payload_data = [
        plc.gpumemoryview(
            rmm.DeviceBuffer.to_device(
                src_bytes,
                plc.utils._get_stream(),
            )
        )
        for r in reader.payload_column_chunks_byte_ranges(
            row_groups, simple_parquet_options
        )
    ]
```

**Review by Matt711 — APPROVED** [review]
Looks good, just one thing todo before merging
